application/actions/authsocial.go

```diff
@@ -0,0 +1,355 @@
+package actions
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/gobuffalo/buffalo"
+	"github.com/gobuffalo/buffalo/render"
+
+	"github.com/silinternational/wecarry-api/auth"
+	"github.com/silinternational/wecarry-api/auth/facebook"
+	"github.com/silinternational/wecarry-api/auth/google"
+	"github.com/silinternational/wecarry-api/auth/linkedin"
+	"github.com/silinternational/wecarry-api/auth/twitter"
+	"github.com/silinternational/wecarry-api/domain"
+	"github.com/silinternational/wecarry-api/models"
+)
+
+const (
+	// http param for social auth provider
+	AuthTypeParam = "auth-type"
+
+	// Auth Type Identifiers
+	AuthTypeFacebook = "facebook"
+	AuthTypeGoogle   = "google"
+	AuthTypeLinkedIn = "linkedin"
+	AuthTypeTwitter  = "twitter"
+
+	envSocialAuthKey    = "key"
+	envSocialAuthSecret = "secret"
+
+	AuthSelectPath = "%s/auth/select/?%s=%s"
+)
+
+// SocialAuthConfig holds the Key and Secret for a social auth provider
+type SocialAuthConfig struct{ Key, Secret string }
+
+// Don't Modify outside of this file.
+var socialAuthConfigs = map[string]SocialAuthConfig{}
+
+// Don't Modify outside of this file.
+var socialAuthSelectors = []authSelector{}
+
+// If there is a Key and Secret, then add them to the auth providers' configs
+// Maps act as pass-by-reference, so configs gets modified in place
+func addConfig(authType, key, secret string, configs map[string]SocialAuthConfig) {
+	if key == "" || secret == "" {
+		return
+	}
+	configs[authType] = SocialAuthConfig{Key: key, Secret: secret}
+}
+
+func addFacebookConfig(configs map[string]SocialAuthConfig) {
+	addConfig(AuthTypeFacebook, domain.Env.FacebookKey, domain.Env.FacebookSecret, configs)
+}
+
+func addGoogleConfig(configs map[string]SocialAuthConfig) {
+	addConfig(AuthTypeGoogle, domain.Env.GoogleKey, domain.Env.GoogleSecret, configs)
+}
+
+func addLinkedInConfig(configs map[string]SocialAuthConfig) {
+	addConfig(AuthTypeLinkedIn, domain.Env.LinkedInKey, domain.Env.LinkedInSecret, configs)
+}
+
+func addTwitterConfig(configs map[string]SocialAuthConfig) {
+	addConfig(AuthTypeTwitter, domain.Env.TwitterKey, domain.Env.TwitterSecret, configs)
+}
+
+// getSocialAuthConfigs returns a map of the enabled Social Auth Provider key-secret pairs
+// In theory, this could go into an init function, but there is already
+// one in render.go and it's a bit risky having multiple init functions.
+func getSocialAuthConfigs() map[string]SocialAuthConfig {
```

Comment:
Would it be possible to do this in an `init()`?

Reply:
It sounds like you didn't read my comment. (-:

Or did you and you still think it should go in one?

Reply:
TBH, I didn't read the comment -- sorry about that. I do think it would still be good to use `init` since the benefits outweigh the risks (which I think are none in this case).

Reply:
OK. How about a hybrid ...
I'll keep modified versions of the two functions for ease of testing and I'll only call them from the init() function.  I'll also give them each a doc comment, "// Forbidden. Do NOT call this function.  (Only called by init() and a test)"

Reply:
Sounds good. Is the "forbidden" warning necessary, since there's a check to see if it's already been called?

Reply:
Well, I removed the check - I thought that was probably why you wanted to have them in the init() function - so we just execute them once.

Reply:
Oh, that makes sense.